### PR TITLE
refactor: initialised `clippy::missing_panics_doc`

### LINKED ISSUE #163: Enable `clippy::missing_panics_doc`
# Background and Motivation

In order to improve documentation and identify possible panics, we should enable clippy lints that warn when a function that panics but does not document this behavior.

# Requested Changes

- [ ] Enable `clippy::missing_panics_doc` lint.
- [ ] Add `#![cfg_attr(test, allow(clippy::missing_panics_doc))]` in order to not warn on `test`-only functions.
- [ ] Add documentation for every function violating this lint. (At the time of writing, this is 43 functions.)
- [ ] Set the `check-private-items = true` option.
- [ ] Add documentation for the private items violating this lint. (At the time of writing, this is 73 _additional_ functions.)

It is ok for this to be done in separate PRs. (e.g. one PR could enable the lint, and simply write `TODO: add panic docs` everywhere the warning appears.
NOTE: `allow(clippy::missing_panics_doc)` should only be used if a `panic` is actually impossible.

### PR DESCRIPTION
**Pull request by Gmin2**
# Rationale for this change

In order to improve documentation and identify possible panics, we should enable clippy lints that warn when a function that panics but does not document this behavior.

# What changes are included in this PR?

Added panic doc comments for all the `unwrap()` statements

Fixes #163 
/claim #163

### CODE REVIEW

**Comment by Gmin2** [comment]
@JayWhite2357 , to complete the doc part of it will take two to three days

File: Cargo.toml
Comment (JayWhite2357) [inline Cargo.toml:71]: NIT:
```suggestion
missing_panics_doc = "warn"

```

File: clippy.toml
Comment (JayWhite2357) [inline clippy.toml:2]: NIT:
```suggestion
check-private-items = true

```

File: crates/proof-of-sql-parser/src/identifier.rs
Comment (JayWhite2357) [inline crates/proof-of-sql-parser/src/identifier.rs:88]: Rather than putting these on each test module, it should work to put
```rust
#![cfg_attr(test, allow(clippy::missing_panics_doc))]
```
at the top of `lib.rs`.

File: crates/proof-of-sql-parser/src/intermediate_ast.rs
Comment (JayWhite2357) [inline crates/proof-of-sql-parser/src/intermediate_ast.rs:228]: Something like this should allow the lints to pass.
```suggestion
    /// Create an `AliasedResultExpr` from an `Expression` using the provided alias.
    ///
    /// # Panics
    ///
    /// TODO: add panic docs
```

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]
If you get this to pass CI, I'm happy to merge this PR while you work on the docs in a separate PR.
It would prevent any new code from being undocumented in the meantime.

**Comment by JayWhite2357** [comment]
Since you've contributed a couple PRs now, it would be helpful for you to be a bit more verbose on the merge requests.
Check out some of the recently merged PRs for some examples: https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/181, https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/180.

**Comment by JayWhite2357** [comment]
@Gmin2 it looks like I can't approve the CI workflows unless the branch conflicts are resolved.

**Comment by Gmin2** [comment]
Hey @JayWhite2357, i think it will make sense to include docs in this PR only, wdyt?

**Comment by JayWhite2357** [comment]
> Hey @JayWhite2357, i think it will make sense to include docs in this PR only, wdyt?

It doesn't matter to me. Entirely up to you.

**Comment by Gmin2** [comment]
Hey @JayWhite2357, completed all the panic docs for `unwrap()` statements, but the assert statements are giving warnings what should i do about them ?

**Comment by JayWhite2357** [comment]
> Hey @JayWhite2357, completed all the panic docs for `unwrap()` statements, but the assert statements are giving warnings what should i do about them ?

@Gmin2 

If the assert is _guaranteed_ to not panic, then it is safe to use `allow(clippy::missing_panics_doc, reason = "some reasoning why we should not include a panic doc")`.
Be sure that the reason is thorough. If needed, put a short reason in the reason = "..." and if a more verbose explanation is needed, put more in some comments (not docs).

File: crates/proof-of-sql-parser/src/identifier.rs
Comment (JayWhite2357) [reply crates/proof-of-sql-parser/src/identifier.rs:93]: I still would like this to be done. Is there a reason we can't do this?

File: crates/proof-of-sql-parser/src/utility.rs
Comment (JayWhite2357) [inline crates/proof-of-sql-parser/src/utility.rs:7]: There is no `schema` here.

File: crates/proof-of-sql-parser/src/utility.rs
Comment (JayWhite2357) [inline crates/proof-of-sql-parser/src/utility.rs:107]:
```suggestion
/// This function will panic if either the `name` or the `schema` (if provided) cannot be parsed as valid [Identifier]s.
```
The reason the "as valid Identifiers" is important is that it is not clear without that clause what would make it parsing not possible.

Please update this in other spots as well.

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:21]: Since this is guaranteed to not happen, it shouldn't be included in the documentation.

Instead, the `unwrap` should really be an `expect` with a reason.

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:23]: This isn't a panic, and is normal behavior. The `0` polynomial is always `0` at every point.

Although if this confused you, it appears that this clarifying note should be added to the documentation (outside of the Panic documentation).

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:18]: I seems as though this should never panic. So, this is a case where `allow`ing the lint is appropriate. (With a `reason =`.)

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:86]: Check out this:
https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/173ec574d93118b5c6344b9ea1713939b26d0481/crates/proof-of-sql/src/base/polynomial/interpolate_test.rs#L98-L163

These cases behave properly without panic.

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:87]: The product should never be zero because the numbers being multiplies are all non-zero.
This should be explained in an `expect` rather than an `unwrap`.

File: crates/proof-of-sql/src/base/polynomial/interpolate.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/interpolate.rs:88]: The `unwrap_or` doesn't panic. It returns `vec![]` if the `unwrap` _would_ have `panic`ed.

File: crates/proof-of-sql/src/base/proof/transcript_core.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/proof/transcript_core.rs:20]: Can this happen? What would cause `read_from` to fail?

File: crates/proof-of-sql/src/base/scalar/mont_scalar.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/scalar/mont_scalar.rs:135]: Take a look at https://docs.rs/ark-ff/latest/ark_ff/fields/trait.PrimeField.html#tymethod.from_bigint.

Explain what "invlid input" means.

File: crates/proof-of-sql/src/base/scalar/mont_scalar.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/scalar/mont_scalar.rs:285]: I'm not sure this failure can occur.

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]
I started reviewing some of this.

The general feedback I have on this: there are some cases where an `unwrap` or another panic _could_ occur, but based on the logic of the methods, doesn't occur. This should not be included in the Panic documentation, and instead should be `expect`ed away and explained in comments (no docs).

Additionally, there are a couple of spots I noted where the panic documentation could be a bit more descriptive on the root cause of _why_ the panic occurs.

In both cases, you may have to look at, for example, _why_ a method returns `None` so that you can explain either _why_ a panic might occur or _why_ it cannot occur.

**Comment by JayWhite2357** [comment]
@Gmin2 thanks for the effort here. It is looking good!

File: crates/proof-of-sql-parser/src/identifier.rs
Comment (Gmin2) [reply crates/proof-of-sql-parser/src/identifier.rs]: done in this [commit](https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/188/commits/af0941d24812df74eed828087e098f2d203b4002)

File: crates/proof-of-sql/src/base/proof/transcript_core.rs
Comment (Gmin2) [reply crates/proof-of-sql/src/base/proof/transcript_core.rs:20]: ig yes

**Comment by Gmin2** [comment]
> @Gmin2 thanks for the effort here. It is looking good!

Hey @JayWhite2357 , i have added all the changes you have mentioned, can you take a look and one more thing can we wrap this quickly because the merge conflicts are getting bigger and bigger, if somethings needs to be adjusted we can do it in the followup PR

**Review by JayWhite2357 — APPROVED** [review]

**Pull request merged**
